model_signing/signature/verifying.py

```diff
@@ -0,0 +1,158 @@
+from cryptography.hazmat.primitives.hashes import SHA256
+from OpenSSL import crypto as ssl_crypto
+from google.protobuf import json_format
+from in_toto_attestation.v1 import statement_pb2 as statement_pb
+from sigstore.verify import verifier as sig_verifier
+from sigstore.verify import policy as sig_policy
+from sigstore.verify import models as sig_models
+from sigstore_protobuf_specs.dev.sigstore.bundle import v1 as bundle_pb
+
+from signature import utils
+
+
+@dataclass
+class VerificationResult:
+    passed: bool
+    information: str = ''
+
+class Verifier(abc.ABC):
+    """Verifier is the abstract base class for all verifying methods."""
+
+    @abc.abstractmethod
+    def verify(self, bundle: bundle_pb.Bundle) -> VerificationResult:
+        """Verify the signature of the provided bundle.
+
+        Args:
+            bundle (bundle_pb.Bundle): the bundle that needs to be verified.
+
+        Returns:
+            VerificationResult: status information about the verification.
+        """
+        pass
+
+class FakeVerifier(Verifier):
+    """Provides a fake verifier that always passes."""
+    def verify(self, bundle: bundle_pb.Bundle) -> VerificationResult:
+        return VerificationResult(passed=True)
+
+class SigstoreVerifier(Verifier):
+    """Provides a verifier using sigstore."""
+
+    def __init__(self, oidc_provider: str, identity: str):
+        self._verifier = sig_verifier.Verifier.production()
+        self._policy = sig_policy.Identity(
+            identity=identity,
+            issuer=oidc_provider,
+        )
+
+    def verify(self, bundle: bundle_pb.Bundle) -> VerificationResult:
+        try:
+            sig_bundle = sig_models.Bundle(bundle)
+            payload = self._verifier.verify_dsse(sig_bundle, self._policy)
+        except Exception as e:
+            return VerificationResult(passed=False, information=str(e))
+        return VerificationResult(passed=True, information='')
+
+class KeyVerifier(Verifier):
+    """Provides a verifier using a public key."""
+
+    def __init__(self, public_key: crypto_ec.EllipticCurvePublicKey):
+        self._public_key = public_key
+    
+    @classmethod
+    def from_path(cls, key_path: str):
+        with open(key_path, 'rb') as fd:
+            serialized_key = fd.read()
+        public_key = crypto_serialization.load_pem_public_key(serialized_key)
+        return cls(public_key)
+
+    def verify(self, bundle: bundle_pb.Bundle) -> VerificationResult:
+        statement = json_format.Parse(bundle.dsse_envelope.payload, statement_pb.Statement())
+        pae = utils.pae(statement)
+        try:
+            self._public_key.verify(bundle.dsse_envelope.signatures[0].sig, pae, crypto_ec.ECDSA(SHA256()))
+        except Exception as e:
+            return VerificationResult(passed=False, information='signature verification failed ' + str(e))
+        return VerificationResult(passed=True)
+        
+
+class PKIVerifier(Verifier):
```

Comment:
It would be useful for users of PKISIgner and verifier to not be forced to pull in Sigstore dependencies. Can we have PKISigner in it own folder `pki`?  to achieve that?

Reply:
Done.